packages/@react-spectrum/s2/src/Tabs.tsx

```diff
@@ -10,9 +10,342 @@
+  children?: ReactNode
+}
+
+interface TabPanelProps extends Omit<AriaTabPanelProps, 'children' | 'style' | 'className'>, UnsafeStyles {  
+  /** Spectrum-defined styles, returned by the `style()` macro. */
+  styles?: StylesPropWithHeight,
+  /** The content to display in the tab panels. */
+  children?: ReactNode
+}
+
+const tabPanel = style({
+  marginTop: 4,
+  color: 'gray-800',
+  flexGrow: 1,
+  flexBasis: '[0%]',
+  minHeight: 0,
+  minWidth: 0
+}, getAllowedOverrides({height: true}));
+
+export function TabPanel(props: TabPanelProps) {
+  return (
+    <AriaTabPanel
+      {...props}
+      style={props.UNSAFE_style}
+      className={(props.UNSAFE_className || '') + tabPanel(null , props.styles)} />
+  );
+}
+
+const tab = style({
+  ...focusRing(),
+  display: 'flex',
+  color: {
+    default: 'neutral-subdued',
+    isSelected: 'neutral',
+    isHovered: 'neutral-subdued',
+    isDisabled: 'disabled',
+    forcedColors: {
+      isSelected: 'Highlight',
+      isDisabled: 'GrayText'
+    }
+  },
+  borderRadius: 'sm',
+  gap: 'text-to-visual',
+  height: {
+    density: {
+      compact: 32,
+      regular: 48
+    }
+  },
+  alignItems: 'center',
+  position: 'relative',
+  cursor: 'default',
+  flexShrink: 0
+}, getAllowedOverrides());
 
-export function Tabs(props: TabsProps) {
-  return <RACTabs {...props} />;
+const icon = style({
+  flexShrink: 0,
+  '--iconPrimary': {
+    type: 'fill',
+    value: 'currentColor'
+  }
+});
+
+export function Tab(props: TabProps) {
+  let {density} = useContext(TabsInternalContext);
+
+  return (
+    <RACTab
+      {...props}
+      style={props.UNSAFE_style}
+      className={renderProps => (props.UNSAFE_className || '') + tab({...renderProps, density}, props.styles)}>
+      <Provider
+        values={[
+          [TextContext, {className: style({order: 1})}],
+          [IconContext, {
+            render: centerBaseline({slot: 'icon', className: style({order: 0})}),
+            styles: icon
+          }]
+        ]}>
+        {typeof props.children === 'string' ? <Text>{props.children}</Text> : props.children}
+      </Provider>
+    </RACTab>
+  );
 }
+
+const tablist = style({
+  display: 'flex',
+  gap: {
+    orientation: {
+      horizontal: {
+        density: {
+          compact: 24,
+          regular: 32
+        }
+      }
+    }
+  },
+  flexDirection: {
+    orientation: {
+      vertical: 'column'
+    }
+  },
+  paddingEnd: {
+    orientation: {
+      vertical: 20
+    }
+  },
+  paddingStart: {
+    orientation: {
+      vertical: 12
+    }
+  },
+  flexShrink: 0,
+  flexBasis: '[0%]'
+});
+
+export function TabList<T extends object>(props: TabListProps<T>) {
+  let {density, isDisabled, disabledKeys, orientation} = useContext(TabsInternalContext);
+  let state = useContext(TabListStateContext);
+  let [selectedTab, setSelectedTab] = useState<HTMLElement | undefined>(undefined);
+  let tablistRef = useRef<HTMLDivElement>(null);
+
+  useLayoutEffect(() => {
```

Comment:
known issue: when you set the orientation to vertical and toggle between the two densities, the selection indicator won't resize properly. 

i took a look at it with rob and we couldn't get to a solution and we also found that it's an issue in v3. it just was never noticeable because we didn't have controls on the stories. 

we tried a couple of things. mainly moving the resize code out of `<TabLine/>` and into `<TabList/>`, updating some the dependencies, using refs instead of a state, removing one of the `<TabLine/>` so that there was only one, etc but none of these fixed the issue.

Reply:
link to the ticket: https://github.com/orgs/adobe/projects/19/views/32?sliceBy%5Bvalue%5D=%F0%9F%8C%88&pane=issue&itemId=74700408

Reply:
Maybe eventually we could implement the [FLIP animation technique](https://css-tricks.com/animating-layouts-with-the-flip-technique/). This would render the line inside the selected tab so it automatically has the right layout for the current tab and we don't need to do any manual syncing. When switching tabs, the tab line inside the newly selected tab gets offset such that it appears visually like the line that was previously inside the old selected tab, and then transitions to its final position. [This video](https://www.youtube.com/watch?v=kep_Iaxuzy0) shows the overall technique. This is easier to implement with a library like Framer Motion but we could probably make it work.